### PR TITLE
Preliminary work to change the 3d party libraries management process

### PR DESCRIPTION
**Pull request by crsib**
To make dependencies easier to maintain and as an essential step towards the dependencies devendoring for Linux, we are changing the way we handle dependencies.

We will be using Conan as a package manager. For packages that are not yet available in `conan.io`, we will use a self-hosted Artifactory instance: https://artifactory.audacityteam.org/. Recipes are available at https://github.com/audacity/conan-recipes.

Updated build documentation is available in [BUILDING.md](BUILDING.md) file.

---

Initially, we've planned only to migrate wxWdigets in the first step.
However, we had to add the following libraries as well:
* `zlib`, as it is a common dependecy for `wxWidgets`, `lame` and `id3tag`.
* `expat`, as `wxWidgets` and `Audacity` depend on it.  
* `lame`, `id3tag`, as they depend on `zlib`.
* `mad`. Well, it just happened.
For now, all the libraries except wxWidgets use `conan.io` packages. `wxWidgets` recipe is based on the Bincrafters but is heavily modified.

In the future, we plan to use custom recipes for `mad` and `id3tag`. The reason is that they use very outdated build scripts, which do not work on AppleSillicon machines even when cross-compiling.

---

We use Docker to construct a clean Ubuntu 20.04 environment with the required packages to verify that we can build using system libraries only. We use this environment to check that Audacity builds with as many system packages as possible.

To define a Conan dependency, we have added `add_conan_lib` CMake function. This function allows to add a dependency using Conan with two possible system fallbacks:

1. pkg_check_modules is invoked if `PGK_CONFIG ...` is present.
2. find_package is invoked if `FIND_PACKAGE_OPTIONS` is present and `pkg_check_modules` has failed.

If `ALWAYS_ALLOW_CONAN_FALLBACK` is present - `audacity_obey_system_dependencies` will be ignored for the package.

### CODE REVIEW

File: linux/build-environment/Dockerfile
Comment (shoogle) [inline linux/build-environment/Dockerfile:8]: I personally think it is neater to put all `RUN` commands in a separate Bash script that gets copied into the Docker image. See this [example](https://github.com/musescore/MuseScore/blob/b30a63ff3822869cf9186bd92e1b9dc6983238a9/build/Linux%2BBSD/portable/x86_64/Dockerfile#L4-L5) for MuseScore, and also note the surrounding comments about how to reduce the image size (and how not to do it). [Here](https://github.com/musescore/MuseScore/blob/b30a63ff3822869cf9186bd92e1b9dc6983238a9/build/Linux%2BBSD/portable/Recipe#L116-L119) is where the tidying up happens.

File: linux/build-environment/Dockerfile
Comment (shoogle) [inline linux/build-environment/Dockerfile:19]: Switching to a Bash script would also allow you to [use arrays](https://github.com/musescore/MuseScore/blob/b30a63ff3822869cf9186bd92e1b9dc6983238a9/build/Linux%2BBSD/portable/Recipe#L48-L68) to avoid endless backslashes. You can also [add comments](https://github.com/musescore/MuseScore/blob/b30a63ff3822869cf9186bd92e1b9dc6983238a9/build/Linux%2BBSD/portable/Recipe#L24) in arrays, unlike with backslashes.

File: linux/build-environment/build_audacity.sh
Comment (shoogle) [inline linux/build-environment/build_audacity.sh:17]: Here's another place when an array would be ideal:

```bash
cmake_options=(
    -G "Unix Makefiles"
    -Daudacity_lib_preference=system
    -Daudacity_obey_system_dependencies=On
    -Daudacity_use_wxwidgets=local # sorry Linux maintainers!
)

cmake "${cmake_options[@]}" ../audacity
```

File: BUILDING.md
Comment (shoogle) [inline BUILDING.md:165]: This table is going to be a pain to maintain in Markdown. I recommend linking to the relevant CMakeLists.txt where the options are defined, or if they are in separate files then just tell people to use the `cmake -LH` option.

If you really want a separate list, I recommend putting it in an online spreadsheet and linking to it from here. I'll PM you with more details about this.

File: .github/workflows/cmake_build.yml
Comment (JamesCrook) [inline .github/workflows/cmake_build.yml:153]: Why pch disabled?

File: BUILDING.md
Comment (shoogle) [inline BUILDING.md:207]: This is more information that would be better in a spreadsheet.

Ideally the CMake configuration would simply fail with an appropriate error message if the correct library version is not available:

```
Error: expat version 2.2.9 or later is required. Please install it with this command:

  $ sudo apt install libexpat1-dev
```

You can give the command or link to the Wiki page that has the command. This [PR for MuseScore](https://github.com/musescore/MuseScore/pull/6262) does something similar if Qt is not installed.

**Review by shoogle — CHANGES_REQUESTED** [review]
I've not used Conan before but it's looks pretty neat!

**Review by JamesCrook — COMMENTED** [review]
Looks way cleaner than before.
Will Artifactory be configured to allow anonymous access?

**Comment by Be-ing** [comment]
>  Will Artifactory be configured to allow anonymous access?

I am wondering this as well. Will developers be able to build with the same builds of libraries locally that the official builds use?

**Comment by Be-ing** [comment]
> In the future, we plan to use custom recipes for mad and id3tag. The reason is that they use very outdated build scripts, which do not work on AppleSillicon machines even when cross-compiling.

These libraries seem to be abandoned for many years. @Holzhaus wrote a [CMakeLists.txt for libid3tag](https://github.com/mixxxdj/vcpkg/blob/2.3/overlay/ports/libid3tag/CMakeLists.txt) which may come in handy for you. Maybe you could fork these old libraries and host them under the Audacity GitHub organization?

File: BUILDING.md
Comment (Be-ing) [inline BUILDING.md:130]:
```suggestion
   $ cmake -Daudacity_use_ffmpeg=loaded ../audacity
```
Unix Makefiles is the default generator on Linux and macOS. There is no need to specify it explicitly (I personally use Ninja though).

File: BUILDING.md
Comment (Be-ing) [inline BUILDING.md:154]: Thank you for documenting these.

File: BUILDING.md
Comment (Be-ing) [inline BUILDING.md:235]:
```suggestion
2. **portaudio-v19**: Audacity currently uses some private APIs (#871), so using system portaudio is not yet possible.
```

File: BUILDING.md
Comment (crsib) [reply BUILDING.md:1]: We had a specific request that there is a manual available in the source code. But I will put it on the wiki page as well after the branch has merged

File: linux/build-environment/Dockerfile
Comment (crsib) [reply linux/build-environment/Dockerfile:8]: Nice catch, I will look into on Monday

File: BUILDING.md
Comment (crsib) [reply BUILDING.md:165]: Yeah, that why I have called it "most notable" LOL

I guess I will move it in a separate doc a rewrite it a bit. Also, VS code makes such tables really easy to format :-)

File: BUILDING.md
Comment (crsib) [reply BUILDING.md:207]: >This is more information that would be better in a spreadsheet.

We have a spreadsheet, but I guess we need an "offline" version as well. I will move this table into a separate doc under the `linux` directory.

File: .github/workflows/cmake_build.yml
Comment (crsib) [reply .github/workflows/cmake_build.yml:153]: That is a good question, I don't know :-) It is disabled for all the CI builds for some reason, hurts Paul sometimes

File: BUILDING.md
Comment (JamesCrook) [reply BUILDING.md:235]: `[some private APIs](https://github.com/audacity/audacity/issues/871)` ?

File: BUILDING.md
Comment (crsib) [reply BUILDING.md:9]: Absolutely not. But it is easier to install CMake using brew.

For Conan - pip3 is always installed with the latest versions of macOS

File: BUILDING.md
Comment (crsib) [reply BUILDING.md:249]: Yeah, I'll move most of it under `linux/minimal_libraries_versions.md`.

We had a request from the maintainers that we keep the build docs in source tree

**Comment by crsib** [comment]
> > Will Artifactory be configured to allow anonymous access?
> 
> I am wondering this as well. Will developers be able to build with the same builds of libraries locally that the official builds use?

@Be-ing @JamesCrook 

It is already available :) The packages list is empty, because of some Artifactory limitations I wasn't able to overcome - it simply doesn't list packages with anonymous access. You can browse the list here: https://artifactory.audacityteam.org/ui/repos/tree/General/conan-local

Or you can run 

```
conan remote add audacity https://artifactory.audacityteam.org/artifactory/api/conan/conan-local
```

and play with wxwidgets build using:

```
conan install wxwidgets/3.1.3-audacity@
```

>Maybe you could fork these old libraries and host them under the Audacity GitHub organization?

We will have to :-) We have important patches in the build tree. Likely I will do that early on Monday

File: BUILDING.md
Comment (crsib) [reply BUILDING.md:235]: An excellent point!

File: BUILDING.md
Comment (JamesCrook) [reply BUILDING.md:249]: That sounds a good balance between 'build docs in source tree' and 'provide alternative/elaborating info outside the main doc'.

File: .github/workflows/cmake_build.yml
Comment (Paul-Licameli) [reply .github/workflows/cmake_build.yml:153]: Building with PCH off requires the insertion of more #include directives to make the build pass, and those extra #includes give the honest and complete picture of dependencies with my scripts.

File: BUILDING.md
Comment (Be-ing) [reply BUILDING.md:1]: I don't think it's a great idea to duplicate the content in multiple places. It is very likely one would not be updated when the other is. However a link from the wiki to this document in the code repository would be a good idea.

Comment (Be-ing) [reply BUILDING.md:1]: Also, I don't recommend using two wikis at once. If you're going to add more to GitHub Wiki, I suggest moving everything from MediaWiki and closing the MediaWiki to edits.

File: BUILDING.md
Comment (Be-ing) [inline BUILDING.md:195]: I see `buildsystem=1` listed in the examples above but it is not documented here. What does that option do?

**Comment by Be-ing** [comment]
So does this branch make CMake automatically download the library binaries from Artifactory? If so, please document that and how to disable it.

**Comment by crsib** [comment]
> So does this branch make CMake automatically download the library binaries from Artifactory? If so, please document that and how to disable it.

@Be-ing you mean you always want to build libraries from source? Ok, that's not a problem.

For package-maintainer-like cases, it is possible to share Conan cache, but given that they will be forcing everything to system libraries, I don't think that it is much of an issue really.

**Comment by Be-ing** [comment]
I don't want to build the entire dependency graph from source. The default behavior I expect from a build system is to first search the system for dependencies. I think this should be the same whether building Audacity locally or packaging it. I also want to be able to use the exact dependencies locally that are used for the official builds, which should be opt in.

The documentation distinguishes between "local" and "system" libraries. I understand what "system" libraries mean, but how does Conan fit into this? Does Conan automatically download builds of "local" libraries?

File: BUILDING.md
Comment (crsib) [reply BUILDING.md:195]: This activates so-called "Legacy Build System" in XCode. Their new system is very confusing and we currently have odd issues with it.

**Comment by crsib** [comment]
> I don't want to build the entire dependency graph from source. The default behavior I expect from a build system is to first search the system for dependencies. I think this should be the same whether building Audacity locally or packaging it. I also want to be able to use the exact dependencies locally that are used for the official builds, which should be opt in.

This behavior can be activated using `-Daudacity_lib_preference=system`. However, as I've mentioned earlier, this can be dangerous because there could be libraries with _official_ recipes from `conan.io` that will link to the Conan provided ZLib (or any other library), and you will have issues with the build. This is why this behavior is opt-in, not opt-out. 

If you turn library preference to **system** and you have the matching library installed and properly configured by the system maintainers the Conan won't be used for the library at all. However, this will generally require you to install all the libraries and patched pkg-config files (if you are running Ubuntu)

> The documentation distinguishes between "local" and "system" libraries. I understand what "system" libraries mean, but how does Conan fit into this? Does Conan automatically download builds of "local" libraries?

For the dependencies, that are using Conan - **local** will use Conan provided library. That means that the recipe for the library will be downloaded from `conan.io` or our Artifactory. If there are cached binaries for your system settings - the cached binaries will be downloaded. Otherwise, the dependency will be built locally on your machine. In both cases, your local cache will be checked first.

File: BUILDING.md
Comment (crsib) [reply BUILDING.md:130]: I'd rather prefer to have it explicit

File: BUILDING.md
Comment (crsib) [reply BUILDING.md:153]: We should, but not in this PR

**Comment by Be-ing** [comment]
I'm still confused. Does `system` use Conan as a fallback if the library is not found on the system?

> For the dependencies, that are using Conan - local will use Conan provided library. That means that the recipe for the library will be downloaded from conan.io or our Artifactory. If there are cached binaries for your system settings - the cached binaries will be downloaded. Otherwise, the dependency will be built locally on your machine. In both cases, your local cache will be checked first.

I suggest renaming `local` to `conan`. To me, `local` implies that no network connections will be made.

**Comment by crsib** [comment]
> I'm still confused. Does `system` use Conan as a fallback if the library is not found on the system?

This is documented in this pull request, but I will further clarify it in the next commit. There is an option, that will force only using the system libraries. By default, if there is no system library - some fallback will be used.

> I suggest renaming `local` to `conan`. To me, `local` implies that no network connections will be made.

This PR does not replace all of the dependencies. This option can be renamed to **vendored**, but this will be confusing as well.

The option will be renamed to Conan when the work on dependencies is complete

**Review by shoogle — APPROVED** [review]

**Pull request merged**
